### ISSUE TITLE
[TW-1157] Logging

### ISSUE THREAD

**Issue by taskwarrior** [GothenburgBitFactory/taskwarrior#1192]
_Paul Beckingham on 2009-07-19T12:25:03Z says:_ 

 Add an optional logging capability, enabled with:

    rc.logging:on

which captures various interactive related events (key presses) and debug messages that would otherwise go to stdout, but in the interactive case cannot, and must therefore go to a log file.  The file would be:

    ~/.task/task.log

The file would only be appended to, so the size of the file would not affect performance, but there should be support for:

    rc.log.file.max.size:1MB

or similar.  once the file hits the size, it gets rotated out.

**Comment by taskwarrior** [comment]
Migrated metadata:
```
Created: 2009-07-19T12:25:03Z
Modified: 2014-02-09T02:38:07Z
```

**Comment by taskwarrior** [comment]
_Peter De Poorter on 2010-11-20T05:50:42Z says:_ 

 It should be very interesting to have a general log for logging of all activity.
This log can be used as base for reporting, auditing or can just be consulted by the user. 

Having a good log can be a prerequisite for the more easy implementation of other feature requests:
*  see http://taskwarrior.org/issues/show/318
*  I had the same idea when working with the stop, start journaling. I wanted to use it to register the time I spent to a task.
  After several stop, starts the annotation list becomes too long and I started thinking how this could be done more generally.
  Creating a log and writing reports on it was the solution direction I was going to.
*  .....

**Comment by taskwarrior** [comment]
_David Patrick on 2010-11-20T11:40:01Z says:_ 

 anything like the personal journal idea ? (see issue #221)

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2010-11-21T11:17:44Z says:_ 

 David's right - #221 is the closest living relative.  This log feature is for debugging.

**Comment by taskwarrior** [comment]
_Peter De Poorter on 2010-11-24T11:27:28Z says:_ 

 I like the journal idea but I prefer a clear separation between logging and journaling. Logging is done automatically by the task program, journaling is done by the user. Logging is not editable by the user, journaling is. Via configuration one could indicate which logging messages are interesting to have automatically in the users journal. Via configuration one also should be able to configure the log level. For example log level 1 logs only the commands issued by the user, log level 3 logs everything.
The logging info is not only interesting for debugging but also for the user to look back at past activity. It could be also used as base for reporting, statistics or answering questions such as in http://taskwarrior.org/issues/show/318. Because a log is not editable by the user it is a fine base for further automatisation.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2010-11-28T11:30:59Z says:_ 

 Note that there are three kinds of logging that may be implemented.  They are:
1.  Task-specific logging.  Given that the undo.data file contains a history of all changes, it would be straightforward to pull all the undo records for a single task and present these as a log of all changes made to a task.  This is already implemented in 1.9.4, as additional information in the 'info' command.
1.  Change logging.  Similar to the above, but in reference to every task.  This data could be presented as a change log for the entire task database.  There are currently no plans to implement this.
1.  General purpose logging.  With a command-line version of task, a log of changes is right there in your shell history, so when describing a problem, you could recall your task commands to reconstruct some scenario.  But with a full-screen, event-driven interactive task, there is no shell history, so the task would need general purpose logging to record keystrokes, actions, warnings, errors etc.  That is what this feature was originally intended to describe.

**Comment by taskwarrior** [comment]
_Paul Beckingham on 2011-03-26T09:56:57Z says:_ 

 No plans to implement.
